src/overlays/actors/ovl_En_River_Sound/z_en_river_sound.c

```diff
@@ -27,10 +26,68 @@ const ActorInit En_River_Sound_InitVars = {
     (ActorFunc)EnRiverSound_Draw,
 };
 
-#endif
+void EnRiverSound_Init(Actor* thisx, GlobalContext* globalCtx) {
+    s32 pad;
+    EnRiverSound* this = THIS;
+    Path* path;
+    s32 pathIndex;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_En_River_Sound/EnRiverSound_Init.s")
+    this->playSound = false;
+    pathIndex = RS_GET_PATH_INDEX(&this->actor);
+    this->actor.params = RS_GET_TYPE(&this->actor);
+    if (pathIndex == 0xFF) {
+        Actor_MarkForDeath(&this->actor);
+        return;
+    }
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_En_River_Sound/EnRiverSound_Update.s")
+    path = &globalCtx->setupPathList[pathIndex];
+    this->pathPoints = Lib_SegmentedToVirtual(path->points);
+    this->numPoints = path->count;
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_En_River_Sound/EnRiverSound_Draw.s")
+void EnRiverSound_Update(Actor* thisx, GlobalContext* globalCtx) {
+    EnRiverSound* this = THIS;
+    Vec3f* worldPos = &this->actor.world.pos;
+    Vec3f eye;
+    s32 bgId;
+
+    Math_Vec3f_Copy(&eye, &globalCtx->view.eye);
+
+    if (this->actor.params < RS_RIVER_DEFAULT_LOW_FREQ) {
+        // All sfx from river_sound that accesses gAudioEnvironmentalSfx is associated with a closed-loop
+        // path that is used to play a regional sfx
+        Actor_GetClosestPosOnPath(this->pathPoints, this->numPoints, &eye, worldPos, true);
+    } else {
+        Actor_GetClosestPosOnPath(this->pathPoints, this->numPoints, &eye, worldPos, false);
+        if (BgCheck_EntityRaycastFloor5(&globalCtx->colCtx, &this->actor.floorPoly, &bgId, &this->actor, worldPos) !=
+            BGCHECK_Y_MIN) {
+            this->soundFreqIndex = SurfaceType_GetConveyorSpeed(&globalCtx->colCtx, this->actor.floorPoly, bgId);
+        } else {
+            this->soundFreqIndex = 0;
+        }
+
+        if (this->soundFreqIndex == 0) {
+            this->soundFreqIndex = this->actor.params - RS_RIVER_DEFAULT_LOW_FREQ;
+        } else {
+            this->soundFreqIndex--;
+        }
+
+        this->soundFreqIndex = CLAMP_MAX(this->soundFreqIndex, 2);
+    }
+}
+
+void EnRiverSound_Draw(Actor* thisx, GlobalContext* globalCtx) {
+    static f32 freqScale[] = {
+        0.7f, // 1 / sqrt(2)
+        1.0f, // 1
+        1.4f, // sqrt(2)
+    };
+    EnRiverSound* this = THIS;
+    s16 params = this->actor.params;
+
+    if (params < RS_RIVER_DEFAULT_LOW_FREQ) {
+        Actor_PlaySfxAtPos(&this->actor, gAudioEnvironmentalSfx[params]);
```

Comment:
I just realized something about this check. `RS_RIVER_DEFAULT_LOW_FREQ` has value 0xFD, so any value lower than that could be used to access `gAudioEnvironmentalSfx`, which only has 0x16 elements.
So, this _could_ be a problem, but I doubt this actor is spawned with the wrong flags on normal conditions (I haven't checked tho).
So, I guess this should be noted as a comment. I'm not sure if it should be labelled as a bug, feels a bit overkill lol

Reply:
I think we decided that indexing an out-of-bounds array isn't enough to be considered a bug as that's up to the user to chose the available params. That's why I included every possible entry in the `RiverSoundType` enum. But I can add a comment to that enum stating that any enum not used in that list is invalid maybe?

Reply:
Yeah, a comment with a small note should be enough